CI/CD Guide (#11443)
PrefectHQ/prefect: docs/guides/ci-cd.md

---
description: CI/CD resources for working with Prefect.
tags:
    - CI/CD
    - continuous integration
    - continuous delivery
search:
  boost: 2
---

# CI/CD with Prefect

> Prefect is used by many organizations in their CI/CD pipelines.

Reviewer: I have a weakly held preference for using the word "process" instead of "pipeline" in order to avoid confusion with most user's association with "data pipeline".

---
description: CI/CD resources for working with Prefect.
tags:
  - CI/CD
  - continuous integration
  - continuous delivery
search:
  boost: 2
---

# CI/CD with Prefect

> Prefect is used by many organizations in their CI/CD pipelines.

Reviewer 2: In most CI/CD discussions involving Prefect, the focus is on deployments. Prefect's deployment features are designed to work well with various CI/CD platforms, enabling efficient management of flow code updates, scheduling changes, and container builds.

> Prefect is used by many organizations in their CI/CD pipelines.

Author: Just rewrote the intro based on your suggestions!

Each organization has their own unique setup, but a common pattern is to use CI/CD to manage Prefect [deployments](/concepts/deployments).
Combining Prefect's deployment features with CI/CD tools enables efficient management of flow code updates, scheduling changes, and container builds.

> This guide utilizes [GitHub Actions](https://docs.github.com/en/actions) to execute CI/CD pipelines, but these concepts are generally applicable across many CI/CD tools.

Reviewer 3: ```suggestion
This guide uses [GitHub Actions](https://docs.github.com/en/actions) to execute CI/CD pipelines, but these concepts are generally applicable across many CI/CD tools.
```

> This guide utilizes [GitHub Actions](https://docs.github.com/en/actions) to execute CI/CD pipelines, but these concepts are generally applicable across many CI/CD tools.

Reviewer: ```suggestion
This guide utilizes [GitHub Actions](https://docs.github.com/en/actions) to implement a CI/CD process, but these concepts are generally applicable across many CI/CD tools.
```

> Note that Prefect's `.deploy` flow method and `prefect.yaml` configuration file are both designed with building and pushing images to a Docker registry in mind.

Reviewer 3: ```suggestion
Note that Prefect's primary ways for creating deployments, a`.deploy` flow method or a `prefect.yaml` configuration file, are both designed with building and pushing images to a Docker registry in mind.
```

## Getting started with GitHub Actions and Prefect

In this example, you'll write a GitHub Actions workflow that will run each time you push to your repository's `main` branch. This workflow will build and push a Docker image containing your flow code to Docker Hub, then deploy the flow to Prefect Cloud.

### Repository secrets

> Your CI/CD pipeline must be able to authenticate to Prefect in order to deploy flows.

Reviewer: ```suggestion
Your CI/CD process must be able to authenticate with Prefect in order to deploy flows.
```

Deploying flows securely and non-interactively in your CI/CD pipeline can be accomplished by saving your `PREFECT_API_URL` and `PREFECT_API_KEY` [as secrets in your repository's settings](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) so they can be accessed in your CI/CD runner's environment without exposing them in any scripts or configuration files.

> Since deploying flows in this scenario also includes building and pushing Docker images, add `DOCKER_USERNAME` and `DOCKER_PASSWORD` as secrets to your repository as well.

Reviewer: ```suggestion
In this scenario, deploying flows involves building and pushing Docker images, so add `DOCKER_USERNAME` and `DOCKER_PASSWORD` as secrets to your repository as well.
```
